File: src/main/scala/com/expedia/gps/geo/reactive101/scala/akka/BasicAkka1_HelloWorld.scala
package com.expedia.gps.geo.reactive101.scala.akka

import akka.actor.{Actor, ActorSystem, Props}

import scala.concurrent.ExecutionContext

/**
  * A basic actor developed in scala.
  */
object BasicAkka1_HelloWorld  {

  val system = ActorSystem("AKKA_test")
  implicit val executionContext: ExecutionContext = system.dispatcher

  def main(args: Array[String]) {
    val myActor = system.actorOf(Props[MyActor])
    myActor ! Hello("Brian")
    system.shutdown()
  }

  case class Hello(name: String)

  class MyActor() extends Actor {
    def receive = {
      case Hello(name) => println(s"Hello $name !")
    }
  }

}

File: src/main/java/com/expedia/gps/geo/reactive101/akka/BasicAkka1_HelloWorld.java
package com.expedia.gps.geo.reactive101.akka;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.actor.UntypedActor;
import lombok.Value;

/**
 * A basic actor developed in Java.
 */
public class BasicAkka1_HelloWorld {

  private static ActorSystem system = ActorSystem.create("AKKA_test_in_Java");

  public static void main(String[] args) {
    ActorRef myActor = system.actorOf(Props.create(MyActor.class));
    myActor.tell(new Hello("Brian"), null);

    try {
      Thread.sleep(100);
    } catch (InterruptedException e) {
      // nothing null
    }
    system.shutdown();
  }


  @Value
  private static class Hello {
    private final String name;
  }

  private static class MyActor extends UntypedActor {

    @Override
    public void onReceive(Object message) throws Exception {
      if (message instanceof Hello) {
        Hello hello = (Hello)message;
        System.out.println("Hello, " + hello.getName() + "!");
      } else
        unhandled(message);
    }
  }
}
